platform/osgi/platform-osgi-configadmin/src/main/java/org/apache/felix/cm/impl/ConfigurationManager.java

```diff
@@ -720,6 +740,28 @@ public void bundleChanged(BundleEvent event) {
 
   // ---------- internal -----------------------------------------------------
 
+  // (CODICE) Extending the proxy to lock on a store means configuration caching is finished
+  // when the lock is released.
+  private static class LockingProxy extends CachingPersistenceManagerProxy {
+    private final StampedLock configDeadZoneLock;
+
+    LockingProxy(final PersistenceManager pm, final StampedLock configDeadZoneLock) {
+      super(pm);
+      this.configDeadZoneLock = configDeadZoneLock;
+    }
+
+    @Override
```

Comment:
❓ How about deletion? Shoudn't we also lock the read lock?

Reply:
👍  

I did not assume that we would be deleting configuration on startup, but since it could hypothetically happen in the future, yes I'll make the change. There is a risk if we finishing acquiring a list of configs, and *then* a config finishes getting deleted, we would resurrect a dead configuration.